src/line.rs

```diff
@@ -347,13 +408,117 @@ impl<'input, Endian> StateMachine<'input, Endian>
+    ///
+    /// The freshly computed row is returned as `Ok(Some((header, row)))`.
+    /// If the matrix is complete, and there are no more new rows in the line
+    /// number matrix, then `Ok(None)` is returned. If there was an error parsing
+    /// an opcode, then `Err(e)` is returned.
+    ///
+    /// Unfortunately, the references mean that this cannot be a
+    /// `FallibleIterator`.
+    pub fn next_row
+        (&mut self)
+         -> parser::Result<Option<(&LineNumberProgramHeader<'input, Endian>, &LineNumberRow)>> {
+        self.inner.next_row()
+    }
+
+    /// Parse and execute opcodes until we reach a row matching `addr`, the end of the program,
+    /// or an error.
+    pub fn run_to_address
+        (&mut self,
+         addr: &u64)
+         -> parser::Result<Option<(&LineNumberProgramHeader<'input, Endian>, &LineNumberRow)>> {
+        self.inner.run_to_address(addr)
+    }
+}
+
+/// Executes a `LineNumberProgram` to recreate the matrix mapping to and from
+/// instructions to source locations.
+///
+/// "The hypothetical machine used by a consumer of the line number information
+/// to expand the byte-coded instruction stream into a matrix of line number
+/// information." -- Section 6.2.1
+///
+/// Like StateMachine, except it can be resumed from sequence boundaries.
+#[derive(Debug, Clone)]
+pub struct ResumableStateMachine<'header, 'input, Endian>
+    where Endian: Endianity + 'input, 'input: 'header
+{
+    inner: StateMachineInner<'input, CompleteLineNumberProgramHeaderHolder<'header, 'input, Endian>, Endian>,
+}
+
+impl<'header, 'input, Endian> ResumableStateMachine<'header, 'input, Endian>
+    where Endian: Endianity
+{
+    /// Get a reference to the header for this state machine's line number
+    /// program.
+    pub fn header(&self) -> &LineNumberProgramHeader<'input, Endian> {
+        self.inner.header.deref()
+    }
+
+    /// Parse and execute the next opcodes in the line number program until
+    /// another row in the line number matrix is computed.
+    ///
+    /// The freshly computed row is returned as `Ok(Some((header, row)))`.
+    /// If the matrix is complete, and there are no more new rows in the line
+    /// number matrix, then `Ok(None)` is returned. If there was an error parsing
+    /// an opcode, then `Err(e)` is returned.
+    ///
+    /// Unfortunately, the references mean that this cannot be a
+    /// `FallibleIterator`.
+    pub fn next_row
+        (&mut self)
+         -> parser::Result<Option<(&LineNumberProgramHeader<'input, Endian>, &LineNumberRow)>> {
+        self.inner.next_row()
+    }
+
+    /// Parse and execute opcodes until we reach a row matching `addr`, the end of the program,
+    /// or an error.
+    pub fn run_to_address
+        (&mut self,
+         addr: &u64)
+         -> parser::Result<Option<(&LineNumberProgramHeader<'input, Endian>, &LineNumberRow)>> {
+        self.inner.run_to_address(addr)
+    }
```

Comment:
Since both `StateMachine` and `ResumableStateMachine` have {header, next_row, run_to_address} methods, it could be nice to define them in a trait and consolidate the documentation in one place.